### ISSUE TITLE
Validation of a new line with required TextFields

### ISSUE THREAD

**Issue by mirkoliver** [TatuLund/GridFastNavigation#69]
Sorry Tatu, I have another issue - I hope, it's the last one. As I don't know, if it's an issue of Vaadin Framework or could also be solved by GridFastNavigation, I'm adressing also you.

It is described here:
[https://github.com/vaadin/framework/issues/10891](https://github.com/vaadin/framework/issues/10891)

I think the reason for this is (from my unaware understanding):
As long, as a cell is not edited, its editor is not validating its value. So if a new row contains null values for required fields, the row can be left with undiscovered errors (as the source of these "errors" is not the editing but the default values of the new row).

Why is this an issue for me?
I'm using a RowEditListener to persist items after succesful edit. In the described case I'm now getting a database error when persisting (insert) the new row, as a required field (constraint in database) isn't filled.

What chances do you see to solve this? Validate all values of a row before an attempt to leave it - and not leaving, when there are validation errors?

**Comment by TatuLund** [comment]
This is a feature or issue (depends on view point) in framework's Binder. Binder assumes that objects getting values that are not input by user, are valid. E.g. from database, or other default values. Hence Validators are run only after fields change. This is not limited to Grid, but Binder behaves like this with Forms as well. It was design decision to implement it this way, so this is somewhat known issue. The opposite implementation would have issues of its own as well.

**Comment by mirkoliver** [comment]
OK, thank you, this makes it clearer.

I think, I've found a way, that almost solves my problem:

`fastNavigation.addRowFocusListener(focusEvent -> getEditor().getBinder().validate());`

This validates all fields of the row when entering its editor. As a result, when there are any required fields that are lacking values, the usual validation errors are displayed and one cannot change the row without filling in all required values.

Why only almost solved? If one leaves the editor via ESC key and then enters it again in the same row, the listener isn't fired again.

Is there a chance to capture this special event?

**Comment by TatuLund** [comment]
Actually this is related to #51

**Comment by TatuLund** [comment]
I modified the code yesterday so that it wont exit the editor by ESC if there are validation errors. I think combination of your idea of triggering validation in row focus event and this small code modification is solution for #51.

**Comment by TatuLund** [comment]
Version 2.2.2 is in Directory

**Comment by mirkoliver** [comment]
Yes, this works much more better! Great!

Unfortunately I have found another special case, in which the FocusEvent is not thrown:

Adding a new line by ...

`items.add(newItem);`
`getDataProvider().refreshAll();`
`select(newItem);`
`getEditor().editRow(items.size()-1);`

... and then clicking on an component outside the grid ...

... then returning and double click on the new line => FocusEvent is not thrown.

**Comment by TatuLund** [comment]
editRow() changes the focus to that row, and once you click out and back, the index of the focused row does not change. At the moment row focus event is thrown each time the index of focused row changes.

**Comment by mirkoliver** [comment]
OK - if I only execute the first three lines (without `editRow`) ...

`items.add(newItem);`
`getDataProvider().refreshAll();`
`select(newItem);`

... it also does not work - even without leaving the grid.

**Comment by TatuLund** [comment]
Reseting the focus tracker is not the way to solve issue, as Grid forces the focus. However based on your original idea I can integrate the force validation into editor opening sequence. IT took a while to solve a timing issue related to that. I have PoC working already, and I need to solve couple of corner cases. I think will make this behavior selectable via API.

**Comment by TatuLund** [comment]
Version 2.2.3 is now in Directory

**Comment by mirkoliver** [comment]
Thank you for this new method! By using it, I don't need to make this on my own any more (by validating on RowFocusEvent). Great!

I've tested your solution more in detail with the following results:

The mentioned problems (see above) are now fixed - row validation works now also in these cases:
- after refreshAll: entering the row editor for the first time
- when editing a row, then clicking on a component outside the grid and returning to the same row in the grid

There are new subsequent problems:
- when being in a column, that has validation error, I cannot leave this column any more without solving the error. It would be better, if I still could navigate within the row from column to column - without the possibility to leave it (the ROW) before fixing all validation errors
- after fixing the validation error in this column I cannot confirm it any more by the tab key. Tab key does not react any more. By mouse click it works.

An additional request:
- leaving a row with validation error by clicking outside the grid is still possible. The user is not informed that the row is unvalid. So one could leave the grid without being informed that there are problems with his edit and that it isn't bound.
To solve this, could the following be accomplished?
    - A row with validation erros cannot be left by clicking outside the grid.
    - The user can cancel this edit by ESC key - the row then is reseted to the values before the edit

**Comment by TatuLund** [comment]
> when being in a column, that has validation error, I cannot leave this column any more without solving the error.

I implemented it this way intentionally, I added in the code check for this. The reason behind this was, that I noticed that Editor has been implemented in the way, that if you edit any cell, the last validation error is cleared, even if you edited different cell that raised the validation error. Is this bug in the Editor, that is separate discussion. Anyway, it is something I cannot alter from the extension. Hence I thought that enforcing to resolve validation errors in the cells which raised it (in left to right) before letting edit other cells, is better approach. 

> leaving a row with validation error by clicking outside the grid is still possible.

Thanks for this one, I forgot to test this use case.

**Comment by mirkoliver** [comment]
- I understand your comment concerning "when being in a column that has validation error, I cannot leave this column any more without solving the error" and I agree, that if this depends on the Editor your solution is the best that can be achieved!
- Concerning "leaving a row with validation error by clicking outside the grid is still possible." - I've tested the fix, but it still doesn't work.
- As a side effect now the row can be left when you only fix one validation error - the other errors in that row are not any more displayed
- Also it would be handy, if the user can cancel his edit by ESC key - the row then is reseted to the values before the edit. But perhaps this idea might be difficult to develop!?

**Comment by TatuLund** [comment]
> Concerning "leaving a row with validation error by clicking outside the grid is still possible." - I've tested the fix, but it still doesn't work.

I re-tested this, and it works for me. Did you remember to recompile the widgetset? Can you describe me how it is not working?

> As a side effect now the row can be left when you only fix one validation error - the other errors in that row are not any more displayed

This I need to investigate more. 

> Also it would be handy, if the user can cancel his edit by ESC key - the row then is reseted to the values before the edit. But perhaps this idea might be difficult to develop!?

Yes, I noticed this regression also. Normally this works, but reseting to value, that does not pass Validation, will not work. Hence this works a bit badly now, since it will not trigger the value change event. So the correct fix I think is to handle ESC key as "save" key in this mode.
